src/Compilers/Core/Portable/Compilation/Compilation.cs

```diff
@@ -1488,6 +1489,161 @@ void addIfNotNull(INamedTypeSymbol? toAdd)
+
+            if (leftType is null)
+                throw new ArgumentNullException(nameof(leftType));
+
+            if (rightType is null)
+                throw new ArgumentNullException(nameof(rightType));
+
+            // When the types are dynamic, then virtually all operators are allowed and synthesized.
+            // So only check when neither are dynamic.
+            if (leftType.TypeKind != TypeKind.Dynamic && rightType.TypeKind != TypeKind.Dynamic)
+            {
+                CheckSupportedBinaryOperatorType(returnType, nameof(returnType));
+                CheckSupportedBinaryOperatorType(leftType, nameof(leftType));
+                CheckSupportedBinaryOperatorType(rightType, nameof(rightType));
+            }
+
+            return CommonCreateBuiltinOperator(name, returnType, leftType, rightType, isChecked);
+
+            void CheckSupportedBinaryOperatorType(ITypeSymbol type, string paramName)
+            {
+                // Enums have operators automatically synthesized for them.
+                if (type.TypeKind == TypeKind.Enum)
+                    return;
+
+                // Pointers have operators automatically synthesized for them.
+                if (type.TypeKind == TypeKind.Pointer)
+                    return;
+
+                // Delegates have a synthesized operator for combining them together.
+                if (type.TypeKind == TypeKind.Delegate)
+                    return;
+
+                switch (type.OriginalDefinition.SpecialType)
+                {
+                    case SpecialType.System_SByte:
+                    case SpecialType.System_Byte:
+                    case SpecialType.System_Int16:
+                    case SpecialType.System_UInt16:
+                    case SpecialType.System_Int32:
+                    case SpecialType.System_UInt32:
+                    case SpecialType.System_Int64:
+                    case SpecialType.System_UInt64:
+                    case SpecialType.System_IntPtr:
+                    case SpecialType.System_UIntPtr:
+                    case SpecialType.System_Char:
+                    case SpecialType.System_Single:
+                    case SpecialType.System_Double:
+                    case SpecialType.System_Decimal:
+                    case SpecialType.System_Boolean:
+                    case SpecialType.System_Object:
+                    case SpecialType.System_String:
+                    case SpecialType.System_DateTime:
+                    case SpecialType.System_Delegate:
+                        return;
+                }
+
+                // ReadOnlySpan<byte> supports a builtin + operator.
+                if (type.OriginalDefinition.Equals(this.CommonGetWellKnownType(WellKnownType.System_ReadOnlySpan_T)?.GetISymbol(), SymbolEqualityComparer.ConsiderEverything) &&
+                    ((INamedTypeSymbol)type).TypeArguments[0].SpecialType == SpecialType.System_Byte)
+                {
+                    return;
```

Comment:
>return; [](http://example.com/codeflow?start=20&length=7)

Is it possible to make this path C# only?

Reply:
it is.  i was on the fence about going to that level or not.  but if you'd prefer that, i'm ok with that.

Reply:
done.